![Full Spectrum Bioinformatics](./cover_image/full_spectrum_bioinformatics_cover.png "A cover image for Full Spectrum Bioinformatics, showing the text title in rainbow colors with a phylogenetic tree, nucleotide substitution diagram, tRNA secondary structure, sequence alignment and principle coordinates analysis plot shown below it.")

This text is currently in early prototype status. Chapters with content you can preview are linked below:

- **Chapter 1. [Foreword](./content/01_foreword/foreword.ipynb)**
- **Chapter 2. Introduction**
- **Chapter 3. The Command Line**
     - **Exercise**:An Adventurous Introduction to the Command Line
- **Chapter 4. Exploring Python**
     - [Exploring Python](./content/04_exploring_python/exploring_python.ipynb)
     - **Exercise**: Error Bingo
     - [A Tour of Python Data Types](./content/04_exploring_python/exploring_python_data_types.ipynb)
     -  A Tour of Python Syntax (functions, conditions, iteration)
- **Chapter 5. Project Design**
     - [Using Literature Surveys to Ask Good Questions and Propose Testable Hypotheses](./content/05_project_design/project_design.ipynb)      
- **Chapter 6. Biological Sequences**
     - [An introduction to Biological Sequences](./content/06_biological_sequences/biological_sequences.ipynb)
     - [Representing and Manipulating Biological Sequences as Python Strings](./content/06_biological_sequences/representing_and_manipulating_biological_sequences_with_python_strings.ipynb)
     - [Analyzing Biological Sequences with For Loops and If Statements](./content/06_biological_sequences/using_for_loops_to_analyze_biological_sequences.ipynb) 
     -  Reading and writing FASTA files using Python
- **Chapter 7. 'Omics**
     -  [An Introduction to 'Omics](./content/07_tabular_omics_data/tabular_omics_data.ipynb) 
     -  **In depth:** Characterizing Microbial Community Ecology
     -  **Exercise:** Microbiome change scenarios
     -  [Working with Tabular 'Omic data in Python using Pandas](./content/07_tabular_omics_data/analyzing_tabular_omics_data_in_pandas.ipynb)    
- **Chapter 8. Alignment and Phylogenetics**
     - **8a. Alignment**
     - Homology and Alignment
     - Global Alignment with the Needleman-Wunsch algorithm
     - Local Alignment with the Smith-Waterman algorithm
     - BLAST and the k-mer trick
     - **8b. Phylogenetics**
     - Tree thinking
     - [Representing Phylogenetic Trees with Python Classes](./content/08_phylogenetic_trees/phylogenetic_trees.ipynb)
     - [Generating Trees Using Birth-Death Models](./content/08_phylogenetic_trees/birth_death_models.ipynb)
     - Working with Traits on Trees
     - Ancestral State Reconstruction
- **Chapter 9. Visualization**
     - Graphs as a Visual Language
     - **Exercise:** Anger Tufte     
- **Chapter 10. Simulation**
     - [Simulating the Population Genetics of Natural Selection and Genetic Drift](./content/10_simulation/simulating_evolution.ipynb)     
- **Chapter 11. Statistics**
     - Monte Carlo simulation and the Funamental Unity of Statistical Hypothesis Tests 
     - Monte Carlo simulation of Effect Size, Sample Size, and Significance 
     - Statistical Distributions
     - Rank Transformations
     - An Introduction to Maximum Likelihood optimization
     - An introduction to Bayesian approaches
- **Chapter 12. Multivariate methods**
     - Of PCoA and Fishtanks     
- **Chapter 13. Presenting Research**
     - Reaching conclusions: 
     - The Inverted Pyramid
- **Chapter 14. Polishing and Publishing**
     - Paper Formatting Guide
     - Paper Formatting Kit 
- **Chapter 15. Careers that draw on Bioinformatics**

- Appendix A - Data Sources for Bioinformatics Projects
- Appendix B - Timesaving Starter Code
     - IUPAC codes in python
     - Standard translation tables in python
- Appendix C - Contributing a Community Example
- Appendix D - Project Specifications 

