{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Representing Traits on Trees\n",
    "\n",
    "Now that we have a handle on phylogenetic trees and an idea of how to represent them, we can add in traits to those trees."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Species can share traits due to homology or convergence\n",
    "\n",
    "When species share a trait due to inheritance from a common ancestor, that is called **homology** and the shared trait is called a **homologous trait**. When species instead share a trait because it evolved more than once on the tree, that is called **convergent evolution** and the shared trait is known as a **homoplasy**. \n",
    "\n",
    "An example of homologous traits vs. convergently evolved traits is shown below:\n",
    "\n",
    "<figure caption-side=\"bottom\">\n",
    "<img src=\"./resources/homology_vs_convergence.png\" width=\"500\" style=\"margin:5px 25px\" description=\"A tree showing a homologous trait arising in an ancestor and being inherited by several of its descendants, and another tree showing the same trait evolving separately in two totally different parts of the tree.\" > \n",
    "<figcaption align=\"left\"><b>Fig. 1</b> - Homologous vs. convergently evolved traits. <br></figcaption>\n",
    "</figure>\n",
    "\n",
    "\n",
    "Whether a particular traits is a homologous trait or a homoplasy depends on the tree. Indeed, modern methods for inferring trees test millions of possible trees to find the ones that do the best job of explaining as many shared traits as possible by homology."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulating Trait Evolution\n",
    "\n",
    "We can add traits to our birth-death model of tree evolution by adding a dict of BinaryTraits to each tree node, and then modifying it as necessary over time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from copy import deepcopy\n",
    "from random import random,choice\n",
    "\n",
    "class PhyloNode():\n",
    "    \"\"\"A node on a phylogenetic tree\"\"\"\n",
    "\n",
    "    def __init__(self,children=None,parent=None,\\\n",
    "      name = None,binary_traits=None):\n",
    "      \"\"\"Initiate a node on a phylogenetic tree\n",
    "      children -- a list of PhyloNode objects\n",
    "        descending immediately from this node\n",
    "      name -- a string for the name of the node\n",
    "      parent -- a single PhyloNode object for the parent\n",
    "      binary_traits -- a dict of True/False traits\n",
    "      \"\"\"\n",
    "      self.Name = name\n",
    "      self.Children = []\n",
    "      if children:\n",
    "        self.Children.extend(children)\n",
    "      self.Parent = parent\n",
    "      self.BinaryTraits = deepcopy(binary_traits) or {}\n",
    "      self.Extinct = False\n",
    "\n",
    "    def isTip(self):\n",
    "      \"\"\"Return True if the node is a tip\"\"\"\n",
    "      if not self.Children: #capture None or []\n",
    "        return True\n",
    "      else:\n",
    "        return False\n",
    "\n",
    "    def isRoot(self):\n",
    "      \"\"\"Return True if the node is the root of the whole tree\"\"\"\n",
    "      if not self.Parent:\n",
    "        return True\n",
    "      else:\n",
    "        return False\n",
    "\n",
    "    def getDescendants(self):\n",
    "        \"\"\"Return a list of PhyloNodes descending from the current node\"\"\"\n",
    "\n",
    "        if self.isTip():\n",
    "            print(self.Name,\" is a tip ... returning []\")\n",
    "            return []\n",
    "\n",
    "        descendants = self.Children or []\n",
    "        for c in self.Children:\n",
    "            #The set of descendants is described\n",
    "            #by the descendants of all the nodes\n",
    "            #immediate children.\n",
    "            if not c.isTip():\n",
    "                child_descendants = c.getDescendants()\n",
    "                descendants.extend([c for c in child_descendants if c not in descendants])\n",
    "\n",
    "            #Side note: this will fail on enormous trees\n",
    "            #due to the recursion limit. Not normally a problem though.\n",
    "        return descendants\n",
    "\n",
    "    def getAncestors(self):\n",
    "        \"\"\"Return the ancestors of the given node\"\"\"\n",
    "        if self.isRoot():\n",
    "            return None\n",
    "\n",
    "        ancestors = [self.Parent]\n",
    "        parents_ancestors = self.Parent.getAncestors()\n",
    "        if parents_ancestors:\n",
    "            ancestors.extend(parents_ancestors)\n",
    "        return ancestors\n",
    "\n",
    "    def getRoot(self):\n",
    "        \"\"\"Return the root node\"\"\"\n",
    "        curr_node = self\n",
    "        while not curr_node.isRoot():\n",
    "            curr_node = curr_node.Parent\n",
    "        return curr_node\n",
    "\n",
    "    def addChild(self,child):\n",
    "        \"\"\"Attach a child node\"\"\"\n",
    "        if child not in self.Children:\n",
    "            self.Children.append(child)\n",
    "        child.Parent = self\n",
    " \n",
    "    def addParent(self,parent):\n",
    "        \"\"\"Attach a parent node\"\"\"\n",
    "        self.Parent = parent\n",
    "        parent.Children.append(self)\n",
    "    \n",
    "    def getMRCA(self,other):\n",
    "        \"\"\"Return PhyloNode for the most recent common ancestor with other\n",
    "        other -- another PhyloNode object in the same tree\n",
    "        \"\"\"\n",
    "        most_recent_common_ancestor = None\n",
    "        #Cache the ancestors of other since\n",
    "        #we'll refer to it often\n",
    "        other_ancestors = other.getAncestors()\n",
    "        self_ancestors = self.getAncestors()\n",
    "        #First check for the trivial case\n",
    "        #where one node is root (but be sure they're on the same tree)\n",
    "        if self.isRoot() and self in other_ancestors:\n",
    "            return self\n",
    "\n",
    "        if other.isRoot() and other in self_ancestors:\n",
    "            return other\n",
    "\n",
    "        for a in self_ancestors:\n",
    "            #Note these will be in order of relatedness\n",
    "            if a in other_ancestors:\n",
    "                #End the loop the first time this happens\n",
    "                #since we want the most\n",
    "                #recent common ancestor\n",
    "                most_recent_common_ancestor = a\n",
    "                break\n",
    "\n",
    "        if not most_recent_common_ancestor:\n",
    "            raise ValueError(\"No common ancestor found for \",self.Name,\" and \",other.Name,\\\n",
    "            \". Are they on the same tree?\")\n",
    "\n",
    "        return most_recent_common_ancestor\n",
    "    \n",
    "    def speciate(self):\n",
    "        \"\"\"Add two children descending from this node\"\"\"\n",
    "        if self.Children:\n",
    "            raise ValueError(\"Internal nodes can't speciate\")\n",
    "        if self.Extinct:\n",
    "            raise ValueError(\"Extinct nodes can't speciate\")\n",
    "\n",
    "        child1_name = self.Name + \"A\"\n",
    "        child1=PhyloNode(name=child1_name,binary_traits=self.BinaryTraits)\n",
    "        self.addChild(child1)\n",
    "        child2_name = self.Name + \"B\"\n",
    "        child2=PhyloNode(name=child2_name,binary_traits=self.BinaryTraits)\n",
    "        self.addChild(child2)\n",
    "\n",
    "\n",
    "    def update(self,speciation_chance=0.25,trait_change_chance=0.25,\\\n",
    "        extinction_chance=0.25):\n",
    "        \"\"\"Update the node by speciation, extinction or trait gain or loss\"\"\"\n",
    "        if not self.isTip():\n",
    "            print(\"Not updating node {name} - it's not a tip\".format(name=self.Name))\n",
    "            return None\n",
    "\n",
    "        if random() < extinction_chance:\n",
    "            print(self.Name,\" goes extinct!\")\n",
    "            self.Extinct = True\n",
    "\n",
    "        if self.Extinct:\n",
    "            print(\"Not updating node {name} - it's extinct\".format(name=self.Name))\n",
    "            return None\n",
    "        print(\"Updating node:{name}\".format(name=self.Name))\n",
    "\n",
    "\n",
    "        for trait,value in self.BinaryTraits.items():\n",
    "            if random() < trait_change_chance:\n",
    "                self.BinaryTraits[trait] = not self.BinaryTraits[trait]\n",
    "                print(self.Name,\" has a new trait value for \",\\\n",
    "                  trait,\": \",self.BinaryTraits[trait],\"!\")\n",
    "\n",
    "        if random() < speciation_chance:\n",
    "            #Speciate\n",
    "            print(\"Node {name} speciates!\".format(name=self.Name))\n",
    "            self.speciate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Root node name: root\n",
      "================================================================================\n",
      "Beginning simulation of time: 1\n",
      "================================================================================\n",
      "About to update node: A\n",
      "Updating node:A\n",
      "A  has a new trait value for  Horn :  True !\n",
      "A  has a new trait value for  Wings :  True !\n",
      "Node A speciates!\n",
      "About to update node: B\n",
      "Updating node:B\n",
      "Nodes in tree at time  1 : 4\n",
      "Tree tips: ['B', 'AA', 'AB']\n",
      "================================================================================\n",
      "Beginning simulation of time: 2\n",
      "================================================================================\n",
      "About to update node: B\n",
      "Updating node:B\n",
      "B  has a new trait value for  Blubber :  True !\n",
      "B  has a new trait value for  Big Ears :  True !\n",
      "About to update node: AA\n",
      "Updating node:AA\n",
      "AA  has a new trait value for  Big Ears :  True !\n",
      "About to update node: AB\n",
      "Updating node:AB\n",
      "AB  has a new trait value for  Red coat :  False !\n",
      "AB  has a new trait value for  Big Ears :  True !\n",
      "AB  has a new trait value for  Horn :  False !\n",
      "AB  has a new trait value for  Wings :  False !\n",
      "Nodes in tree at time  2 : 4\n",
      "Tree tips: ['B', 'AA', 'AB']\n",
      "================================================================================\n",
      "Beginning simulation of time: 3\n",
      "================================================================================\n",
      "About to update node: B\n",
      "Updating node:B\n",
      "B  has a new trait value for  Red coat :  False !\n",
      "Node B speciates!\n",
      "About to update node: AA\n",
      "Updating node:AA\n",
      "AA  has a new trait value for  Wings :  False !\n",
      "Node AA speciates!\n",
      "About to update node: AB\n",
      "AB  goes extinct!\n",
      "Not updating node AB - it's extinct\n",
      "Nodes in tree at time  3 : 8\n",
      "Tree tips: ['AB', 'AAA', 'AAB', 'BA', 'BB']\n",
      "================================================================================\n",
      "Beginning simulation of time: 4\n",
      "================================================================================\n",
      "About to update node: AB\n",
      "AB  goes extinct!\n",
      "Not updating node AB - it's extinct\n",
      "About to update node: AAA\n",
      "AAA  goes extinct!\n",
      "Not updating node AAA - it's extinct\n",
      "About to update node: AAB\n",
      "AAB  goes extinct!\n",
      "Not updating node AAB - it's extinct\n",
      "About to update node: BA\n",
      "Updating node:BA\n",
      "Node BA speciates!\n",
      "About to update node: BB\n",
      "Updating node:BB\n",
      "Node BB speciates!\n",
      "Nodes in tree at time  4 : 12\n",
      "Tree tips: ['AB', 'AAA', 'AAB', 'BAA', 'BAB', 'BBA', 'BBB']\n"
     ]
    }
   ],
   "source": [
    "#Define traits\n",
    "binary_traits = {\"Blubber\":False,\"Red coat\":True,\"Big Ears\":False,\\\n",
    "  \"Horn\":False,\"Wings\":False}\n",
    "#Build the tree!\n",
    "root = PhyloNode(name=\"root\",binary_traits=binary_traits)\n",
    "A = PhyloNode(name=\"A\",binary_traits=binary_traits)\n",
    "B = PhyloNode(name=\"B\",binary_traits=binary_traits)\n",
    "root.addChild(A)\n",
    "root.addChild(B)\n",
    "\n",
    "time = 5\n",
    "#Demo our tree functions\n",
    "print(\"Root node name:\",root.Name)\n",
    "\n",
    "for t in range(1,time):\n",
    "    print(\"=\"*80)\n",
    "    print(\"Beginning simulation of time:\",t)\n",
    "    print(\"=\"*80)\n",
    "    #I want to keep the original set of nodes to update\n",
    "    #so that new species don't immediately get simulated\n",
    "    nodes_to_update = [n for n in root.getDescendants()]\n",
    "    for node in nodes_to_update:\n",
    "\n",
    "        if not node.isTip():\n",
    "            continue\n",
    "        print(\"About to update node:\",node.Name)\n",
    "        node.update()\n",
    "    print(\"Nodes in tree at time \",t,\":\",len([n.Name for n in root.getDescendants()]))\n",
    "    print(\"Tree tips:\",[n.Name for n in root.getDescendants() if n.isTip()])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ancestral State Reconstruction\n",
    "\n",
    "The goal of ancestral state reconstruction (ASR) is to determine the traits of ancestral organisms by comparative analysis of modern organisms. More formally, given annotations for the traits of tip nodes, ancestral state reconstruction attempts to reconstruct the traits at internal nodes.\n",
    "\n",
    "\n",
    "#### The Fitch Parsimony algorithm for Ancestral State Reconstruction."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Molecular Clock Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Molecular clock analysis uses the rate at which traits change over time as a sort of clock. For example, we might expect that if we compare the same snippet of DNA across many species, a pair of species that are 99% similar in DNA sequence in that gene probably diverged much more recently than a pair that are only 76% similar. However, such methods must be careful to account for several factors. \n",
    "\n",
    "First, some changes in DNA sequence will be hidden. For example, imagine a given nucleotide mutated from A to T and then, some years later, back from T to A. This was two changes in the DNA sequence, but a later comparative analysis would see 0 changes! For similar reasons, even two random sequences with equal base frequencies will appear to share 25% sequence identity. \n",
    "\n",
    "A second point where caution is warrented is that the rate of evolution may change in different parts of the tree. For example, if a mutation renders the proof-reading machinery that tries to correct mutations more or less strict, it can decrease or increase the mutation rate in that part of the tree. **Relaxed molecular clocks** try to account for this by allowing different rates of evolution to occur in different parts of the tree."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO:\n",
    "\n",
    "\n",
    "#### Different styles of phylogenetic tree\n",
    "\n",
    "\n",
    "#### The definition of relatedness on a phylogenetic tree\n",
    "\n",
    "The key piece of information that trees convey is the *phylogenetic relatedness* of different organisms. Evolutionary biologists define relatedness in relative terms based on recency of common ancestry. Two organisms A and B are *more related* to one another than they are to another organism C if they share a *more recent common ancestor* with one another than either does to C.  This is a somewhat long definition, but in practice it allows relative relatedness to be tested using a simple procedure.\n",
    "\n",
    "#### Finding the most recent common ancestor (MRCA) of two species by hand\n",
    "To find the common ancestor of A and B, place your fingers on those two nodes on the tree. Then trace backwards along the tree from each to their ancestors. Stop when you hit a node that is an ancestor of both A and B (e.g. the first node that both fingers touch when tracing backwards in the tree from A and B. Mark that node as the most recent common ancestor (MRCA) of A and B.\n",
    "\n",
    "##### A simple test for relatedness on a tree. \n",
    "This test compares whether some species A is more related to B or C. Start with a tree that has on it A, B and C. Trace back along the tree using one's fingers, starting at A and B. Stop when you hit a node that is an ancestor of both A and B (e.g. the first node that both fingers touch when tracing backwards in the tree from A and B. Mark that node as the *most recent common ancestor (MRCA) of A and B*. Now repeat the procedure for A and C and mark that node as *the most recent common ancestor (MRCA) of A and C*. The MRCA of A and B will be one of three things:\n",
    "\n",
    "- the same node as the MRCA of A and C. In that case AB and AC are equally related.\n",
    "- a descendant of the MRCA of A and C. In that case A and B have a more recent common ancestor and therefore A and B are more closely related than A and C\n",
    "- an ancestor of the MRCA of A and C. In that case A and C have a more recent common ancestor and therefore A and C are more closely related than A and B."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Monophyly and Tree Testing\n",
    "\n",
    "##### The Snip Test\n",
    "Have some dogs become unicellular obligate parasites?\n",
    "\n",
    "Hypothesis testing using trees: Canine Transmissible Venereal Tumors \n",
    "\n",
    "##### Tracing zoonotic pathogen outbreaks\n",
    "\n",
    "Zoonotic transmission occurs when a pathogen moves between another species into human hosts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
